File: app/src/main/java/com/waveneuro/ui/base/BaseActivity.java
package com.waveneuro.ui.base;

import android.os.Bundle;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.asif.abase.base.APIStatusView;
import com.asif.abase.base.ProgressWaitView;
import com.asif.abase.data.model.APIError;
import com.asif.abase.data.model.BaseError;
import com.asif.abase.data.model.BaseModel;
import com.messagebar.messagebar.Flashbar;
import com.waveneuro.R;
import com.waveneuro.WaveNeuroApplication;
import com.waveneuro.injection.component.ActivityComponent;
import com.waveneuro.injection.component.DaggerPersistComponent;
import com.waveneuro.injection.component.PersistComponent;
import com.waveneuro.injection.module.ActivityModule;
import com.waveneuro.views.WaveProgressDialog;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;

import timber.log.Timber;

public class BaseActivity extends AppCompatActivity implements ProgressWaitView, APIStatusView {

    private static final String KEY_ACTIVITY_ID = "KEY_ACTIVITY_ID";
    private static final AtomicLong NEXT_ID = new AtomicLong(0);

    private static final Map<Long, PersistComponent> sComponentsMap = new HashMap();
    private ActivityComponent mActivityComponent;
    private WaveProgressDialog waveProgressDialog;
    private long mActivityId;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.mActivityId = savedInstanceState != null ?
                savedInstanceState.getLong(KEY_ACTIVITY_ID) : NEXT_ID.getAndIncrement();



        mActivityComponent = getActivityComponent();
        waveProgressDialog = new WaveProgressDialog(this);

    }

    private ActivityComponent getActivityComponent() {
        PersistComponent persistComponent = null;
        if (!sComponentsMap.containsKey(mActivityId)) {
            Timber.i("Creating new ConfigPersistentComponent id=%d", mActivityId);
            persistComponent = DaggerPersistComponent.builder()
                    .applicationComponent(((WaveNeuroApplication) getApplication()).appComponent)
                    .build();
            sComponentsMap.put(mActivityId, persistComponent);
        } else {
            Timber.i("Reusing ConfigPersistentComponent id=%d", mActivityId);
            persistComponent = sComponentsMap.get(mActivityId);
        }
       return persistComponent.activityComponent(new ActivityModule(this));
    }

    public ActivityComponent activityComponent() {
        if(this.mActivityComponent == null)
            this.mActivityComponent = getActivityComponent();
        return this.mActivityComponent;
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putLong(KEY_ACTIVITY_ID, mActivityId);
    }

    @Override
    protected void onDestroy() {
        if (!isChangingConfigurations()) {
            Timber.i("Clearing ConfigPersistentComponent id=%d", mActivityId);
            sComponentsMap.remove(mActivityId);
        }
        super.onDestroy();
    }

    @Override
    public void onSuccess(BaseModel model) {
        removeWait();
    }

    @Override
    public void onSuccess(List<? extends BaseModel> model) {
        removeWait();
    }

    @Override
    public void onFailure(BaseError error) {
        APIError apiError = (APIError) error;
        String message = "Please contact admin at support@waveneurohelp.zendesk.com";
        if (!TextUtils.isEmpty(apiError.getMessage())) {
            message = apiError.getMessage();
        } else if (!TextUtils.isEmpty(apiError.getMsg())) {
            message = apiError.getMsg();
        } else if (!TextUtils.isEmpty(apiError.getError())) {
            message = apiError.getError();
        }

        removeWait();
        new Flashbar.Builder(BaseActivity.this)
                .gravity(Flashbar.Gravity.BOTTOM)
                .message(message)
                .messageColorRes(R.color.white)
                .backgroundColorRes(R.color.black)
                .primaryActionText("DISMISS")
                .primaryActionTapListener(Flashbar::dismiss)
                .showOverlay()
                .dismissOnTapOutside()
                .build().show();
    }

    @Override
    public void displayWait() {
        if (waveProgressDialog != null && !isFinishing()) {
            waveProgressDialog.startProgressDialog();
        }
    }

    public void displayWait(String title) {
        displayWait(title, null);
    }

    public void displayWait(String title, String message) {
        if (waveProgressDialog != null && !isFinishing()) {
            waveProgressDialog.startProgressDialog();
        }
    }

    @Override
    public void removeWait() {
        if (waveProgressDialog != null && waveProgressDialog.isShowing()) {
            waveProgressDialog.dismiss();
        }
    }

}

File: app/src/main/java/com/waveneuro/ui/base/BaseActivity.kt
package com.waveneuro.ui.base

import android.os.Bundle
import android.text.TextUtils
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.asif.abase.base.APIStatusView
import com.asif.abase.base.ProgressWaitView
import com.asif.abase.data.model.APIError
import com.asif.abase.data.model.BaseError
import com.asif.abase.data.model.BaseModel
import com.waveneuro.R
import com.waveneuro.WaveNeuroApplication
import com.waveneuro.injection.component.ActivityComponent
import com.waveneuro.injection.component.DaggerPersistComponent
import com.waveneuro.injection.component.PersistComponent
import com.waveneuro.injection.module.ActivityModule
import com.waveneuro.views.WaveProgressDialog
import timber.log.Timber
import java.util.concurrent.atomic.AtomicLong

open class BaseActivity : AppCompatActivity(), ProgressWaitView, APIStatusView {

    private var mActivityComponent: ActivityComponent? = null
    private var waveProgressDialog: WaveProgressDialog? = null
    private var mActivityId: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mActivityId = savedInstanceState?.getLong(KEY_ACTIVITY_ID)
            ?: NEXT_ID.getAndIncrement()
        mActivityComponent = getActivityComponent()
        waveProgressDialog = WaveProgressDialog(this)
    }

    private fun getActivityComponent(): ActivityComponent {
        val persistComponent: PersistComponent
        if (sComponentsMap.containsKey(mActivityId)) {
            Timber.i("Reusing ConfigPersistentComponent id=%d", mActivityId)
            persistComponent = sComponentsMap[mActivityId]!!
        } else {
            Timber.i("Creating new ConfigPersistentComponent id=%d", mActivityId)
            persistComponent = DaggerPersistComponent.builder()
                .applicationComponent((application as WaveNeuroApplication).appComponent)
                .build()
            sComponentsMap[mActivityId] = persistComponent
        }
        return persistComponent.activityComponent(ActivityModule(this))
    }

    fun activityComponent(): ActivityComponent? {
        if (mActivityComponent == null)
            mActivityComponent = getActivityComponent()
        return mActivityComponent
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong(KEY_ACTIVITY_ID, mActivityId)
    }

    override fun onDestroy() {
        if (!isChangingConfigurations) {
            Timber.i("Clearing ConfigPersistentComponent id=%d", mActivityId)
            sComponentsMap.remove(mActivityId)
        }
        super.onDestroy()
    }

    override fun onSuccess(model: BaseModel) {
        removeWait()
    }

    override fun onSuccess(model: List<BaseModel>) {
        removeWait()
    }

    override fun onFailure(error: BaseError) {
        val apiError = error as APIError
        var message: String? = getString(R.string.default_error_message)
        if (!TextUtils.isEmpty(apiError.message)) {
            message = apiError.message
        } else if (!TextUtils.isEmpty(apiError.msg)) {
            message = apiError.msg
        } else if (!TextUtils.isEmpty(apiError.error)) {
            message = apiError.error
        }
        removeWait()
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    override fun displayWait() {
        if (waveProgressDialog != null && !isFinishing) {
            waveProgressDialog?.startProgressDialog()
        }
    }

    @JvmOverloads
    fun displayWait(title: String?, message: String? = null) {
        if (waveProgressDialog != null && !isFinishing) {
            waveProgressDialog?.startProgressDialog()
        }
    }

    override fun removeWait() {
        if (waveProgressDialog != null && waveProgressDialog!!.isShowing) {
            waveProgressDialog?.dismiss()
        }
    }

    companion object {
        private const val KEY_ACTIVITY_ID = "KEY_ACTIVITY_ID"
        private val NEXT_ID = AtomicLong(0)
        private val sComponentsMap = mutableMapOf<Long, PersistComponent>()
    }
}